### PR TITLE
Make not passing use_reentrant back to warning instead of erroring and clarify docs

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116710
* #116523

### CODE REVIEW

**Comment by stas00** [comment]
Thank you for working on providing more clarity around this transition, @soulitzer 

This is being discussed at https://pytorch.slack.com/archives/C3PDTEV8E/p1704308146846339

Repasting 2 items from that thread here:

I think the current proposal of this PR is a good start but more is needed: 

1. The discussion of the defaults needs to be updated
2. The current error message isn't very coherent either:
> ValueError: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. [...]

If I try to parse this statement to me it says " The default value will be updated" which implies that there is a default. Which implies it shouldn't be required to pass.

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -372,7 +372,8 @@ def checkpoint(
 
     .. warning::
 
-        You must pass in the ``use_reentrant`` parameter explicitly.
+        The ``use_reentrant`` parameter should be passed explicitly. In a future
+        version of PyTorch, we will error if use_reentrant is not passed.
```
Comment (stas00) [inline torch/utils/checkpoint.py:376]:
```suggestion
        version of PyTorch, we will error if ``use_reentrant`` is not passed.
```

File: torch/utils/checkpoint.py [hunk 2]
```diff
@@ -428,14 +429,16 @@ def checkpoint(
...
             Default: ``True``
-        use_reentrant(bool): Use checkpointing
-            implementation that requires re-entrant autograd.
-            If ``use_reentrant=False`` is specified, ``checkpoint`` will use an
-            implementation that does not require re-entrant autograd. This
+        use_reentrant(bool):
+            specify whether to use the activation checkpoint variant that
+            requires reentrant autograd. This parameter should be passed
+            explicitly. In a future version of PyTorch, we will error if
+            use_reentrant is not passed.
```
Comment (stas00) [inline torch/utils/checkpoint.py:436]:
```suggestion
            ``use_reentrant`` is not passed.
```

File: torch/utils/checkpoint.py [hunk 3]
```diff
@@ -458,14 +461,16 @@ def checkpoint(
...
-        raise ValueError(
-            "torch.utils.checkpoint: please pass in use_reentrant=True or "
-            "use_reentrant=False explicitly. The default value of use_reentrant "
-            "will be updated to be False in the future. To maintain current "
-            "behavior, pass use_reentrant=True. It is recommended that you use "
-            "use_reentrant=False. Refer to docs for more details on the "
-            "differences between the two variants."
+        warnings.warn(
+            "torch.utils.checkpoint: the use_reentrant parameter should be "
+            "passed explicitly. In a future version of PyTorch we will error "
```
Comment (stas00) [inline torch/utils/checkpoint.py:466]: if it's a "version" and not some future version, perhaps say explicitly which version?

File: torch/utils/checkpoint.py [hunk 4]
```diff
@@ -458,14 +461,16 @@ def checkpoint(
...
+            "if use_reentrant is not passed. use_reentrant=False is "
+            "recommended, but if you need to preserve the existing behavior, "
```
Comment (stas00) [inline torch/utils/checkpoint.py:468]:
```suggestion
            "recommended, but if you need to preserve the current default behavior, "
```

File: torch/utils/checkpoint.py [hunk 5]
```diff
@@ -504,9 +509,10 @@ def checkpoint_sequential(functions, segments, input, use_reentrant=None, **kwar
     be saved for re-running the segment in the backward pass.
 
     .. warning::
-        You must pass in the ``use_reentrant`` parameter explicitly.
-        If you are using the ``use_reentrant=True` variant (this is the
-        default), please see :func:`~torch.utils.checkpoint.checkpoint` for
+        The ``use_reentrant`` parameter should be passed explicitly. In a future
+        version of PyTorch, we will error if use_reentrant is not passed.
```
Comment (stas00) [inline torch/utils/checkpoint.py:513]:
```suggestion
        version of PyTorch, we will error if ``use_reentrant`` is not passed.
```

File: torch/utils/checkpoint.py [hunk 6]
```diff
@@ -522,10 +528,13 @@ def checkpoint_sequential(functions, segments, input, use_reentrant=None, **kwar
...
             Default: ``True``
-        use_reentrant(bool): Use checkpointing
-            implementation that requires re-entrant autograd.
-            If ``use_reentrant=False`` is specified, ``checkpoint`` will use an
-            implementation that does not require re-entrant autograd. This
+        use_reentrant(bool):
+            specify whether to use the activation checkpoint variant that
+            requires reentrant autograd. This parameter should be passed
+            explicitly. In a future version of PyTorch, we will error if
+            use_reentrant is not passed.
```
Comment (stas00) [inline torch/utils/checkpoint.py:535]: same comment as above wrt version

**Review by stas00 — COMMENTED** [review]
This version looks excellent, Jeffrey!

I made a few small style consistency tweaks and added a question.

**Comment by soulitzer** [comment]
Updated!

File: torch/utils/checkpoint.py [hunk 7]
```diff
@@ -520,10 +527,12 @@ def checkpoint_sequential(functions, segments, input, use_reentrant=None, **kwar
...
             the RNG state during each checkpoint.
             Default: ``True``
-        use_reentrant(bool): Use checkpointing
-            implementation that requires re-entrant autograd.
-            If ``use_reentrant=False`` is specified, ``checkpoint`` will use an
-            implementation that does not require re-entrant autograd. This
+        use_reentrant(bool):
+            specify whether to use the activation checkpoint variant that
+            requires reentrant autograd. This parameter should be passed
+            explicitly. In version 2.4, we will error if use_reentrant is not
```
Comment (stas00) [inline torch/utils/checkpoint.py:533]:
```suggestion
            explicitly. In version 2.4, we will error if ``use_reentrant`` is not
```

File: torch/utils/checkpoint.py [hunk 8]
```diff
@@ -503,8 +508,10 @@ def checkpoint_sequential(functions, segments, input, use_reentrant=None, **kwar
     be saved for re-running the segment in the backward pass.
 
     .. warning::
-        If you are using the ``use_reentrant=True` variant (this is the
-        default), please see :func:`~torch.utils.checkpoint.checkpoint` for
+        The ``use_reentrant`` parameter should be passed explicitly. In version
+        2.4, we will error if ``use_reentrant`` is not passed.
```
Comment (stas00) [inline torch/utils/checkpoint.py:512]: here and elsewhere perhaps `s/will error/will err/` since `error` is not a verb

Comment (stas00) [reply torch/utils/checkpoint.py:512]: or `throw an exception` or `will assert`

**Review by stas00 — COMMENTED** [review]
That's much better with the actual version! it's a call for action

added a few more small nits

File: torch/utils/checkpoint.py (hunk 8 above)
Comment (stas00) [reply torch/utils/checkpoint.py:512]: ah, and no need for comma, after 2.4

File: test/test_autograd.py [hunk 9]
```diff
@@ -5972,25 +5972,44 @@ def context_fn():
...
             out = checkpoint(lambda x: x.sin(), x, use_reentrant=True, context_fn=context_fn)
 
-    def test_checkpoint_errors_if_use_reentrant_not_passed_explicitly(self):
+    def test_checkpoint_warns_if_use_reentrant_not_passed_explcitly(self):
         a = torch.randn(1, requires_grad=True)
-        error_str = "please pass in use_reentrant=True or use_reentrant=False explicitly"
 
-        with self.assertRaisesRegex(ValueError, error_str):
+        # Passing explicitly should not warn
+        with warnings.catch_warnings(record=True) as w:
```
Comment (albanD) [inline test/test_autograd.py:5979]: with self.assertNotWarn():

File: test/test_autograd.py [hunk 10]
```diff
@@ -5972,25 +5972,44 @@ def context_fn():
...
+            checkpoint(lambda x: x, a, use_reentrant=False)
+        self.assertEqual(len(w), 0)
+
+        # Not passing explicitly warns
+        with warnings.catch_warnings(record=True) as w:
```
Comment (albanD) [inline test/test_autograd.py:5984]: with self.assertWarnsOnceRegex(...):

**Review by albanD — APPROVED** [review]
Sounds good to me as well!
Thanks @stas00 for helping review this!

**Pull request closed without merging**